src/Native/Bootstrap/main.cpp

```diff
@@ -288,16 +288,69 @@ static const pfn c_classlibFunctions[] = {
 
 #endif // !CPPCODEGEN
 
+static bool RUNTIME_INITIALIZED = false;
 extern "C" void InitializeModules(void* osModule, void ** modules, int count, void ** pClasslibFunctions, int nClasslibFunctions);
+extern "C" int InitializeRuntime();
 
+#ifdef CORERT_DLL
 #if defined(_WIN32)
+#define wmain CoreRT_wmain
+#else
+#define main CoreRT_main
+#endif
+#endif // CORERT_DLL
+
+#if defined(_WIN32)
+#ifndef CORERT_DLL
 extern "C" int __managed__Main(int argc, wchar_t* argv[]);
+#else
+extern "C" int __managed__Startup(int argc, wchar_t* argv[]);
+#endif // CORERT_DLL
 int __cdecl wmain(int argc, wchar_t* argv[])
 #else
+#ifndef CORERT_DLL
 extern "C" int __managed__Main(int argc, char* argv[]);
+#else
+extern "C" int __managed__Startup(int argc, char* argv[]);
+#endif // CORERT_DLL
 int main(int argc, char* argv[])
 #endif
 {
+    int initval = InitializeRuntime();
+    if (initval != 0)
+        return initval;
+
+    int retval;
+#ifdef CPPCODEGEN
+    try
+#endif
+    {
+#ifndef CORERT_DLL
+        retval = __managed__Main(argc, argv);
+#else
+        retval = 0;
+#endif // CORERT_DLL
+    }
+#ifdef CPPCODEGEN
+    catch (const char* &e)
+    {
+        printf("Call to an unimplemented runtime method; execution cannot continue.\n");
+        printf("Method: %s\n", e);
+        retval = -1;
+    }
+#endif
+
+    RhpShutdown();
+    RUNTIME_INITIALIZED = false;
+
+    return retval;
+}
+
+extern "C" int InitializeRuntime()
+{
+    if (RUNTIME_INITIALIZED)
```

Comment:
This should be thread safe. If one thread is initializing the runtime, the other thread should wait for it.